cpp/src/hash/hashing.cu

```diff
@@ -77,84 +82,394 @@ class bitwise_partitioner
+
+    // Record the size of this partition in this block
+    const size_type write_location = partition_number * gridDim.x + blockIdx.x;
+    block_partition_sizes[write_location] = block_partition_size;
+    partition_number += blockDim.x;
+  }
+}
+
+/* --------------------------------------------------------------------------*/
+/** 
+ * @brief Move one column from the input table to the hashed table.
+ * 
+ * @param[in] input_buf Data buffer of the column in the input table
+ * @param[out] output_buf Preallocated data buffer of the column in the output table
+ * @param[in] num_rows The number of rows in each column
+ * @param[in] num_partitions The number of partitions to divide the rows into
+ * @param[in] row_partition_numbers Array that holds which partition each row belongs to
+ * @param[in] row_partition_offset Array that holds the offset of each row in its partition of
+ * the thread block.
+ * @param[in] block_partition_sizes Array that holds the size of each partition for each block
+ * @param[in] scanned_block_partition_sizes The scan of block_partition_sizes
+ */
+/* ----------------------------------------------------------------------------*/
+template <typename DataType>
+__global__
+void move_to_output_buffer(DataType const *input_buf,
+                           DataType *output_buf,
+                           const size_type num_rows,
+                           const size_type num_partitions,
+                           size_type * row_partition_numbers,
+                           size_type * row_partition_offset,
+                           size_type * block_partition_sizes,
+                           size_type * scanned_block_partition_sizes)
+{
+  extern __shared__ char shared_memory[];
+  DataType *block_output = (DataType *)shared_memory;
+  size_type *partition_offset_shared = (size_type *)(block_output + BLOCK_SIZE * ROWS_PER_THREAD);
+  size_type *partition_offset_global = (size_type *)(partition_offset_shared + num_partitions + 1);
+
+  size_type ipartition;
+
+  typedef cub::BlockScan<size_type, BLOCK_SIZE> BlockScan;
+  __shared__ typename BlockScan::TempStorage temp_storage;
+
+  // use ELEMENT_PER_BLOCK=2 to support upto 1024 partitions 
+  size_type temp_histo[ELEMENT_PER_BLOCK];
+
+  for (int i = 0; i < ELEMENT_PER_BLOCK; ++i) {
+    if (ELEMENT_PER_BLOCK * threadIdx.x + i < num_partitions) {
+      temp_histo[i] = block_partition_sizes[blockIdx.x + (ELEMENT_PER_BLOCK * threadIdx.x + i) * gridDim.x]; 
+    } else {
+      temp_histo[i] = 0;
+    }
+  }
+
+  __syncthreads();
+
+  BlockScan(temp_storage).InclusiveSum(temp_histo, temp_histo);
+
+  __syncthreads();
+
+  if (threadIdx.x == 0) {
+    partition_offset_shared[0] = 0;
+  }
+
+  for (int i = 0; i < ELEMENT_PER_BLOCK; ++i) {
+    if (ELEMENT_PER_BLOCK * threadIdx.x + i < num_partitions) {
+      partition_offset_shared[ELEMENT_PER_BLOCK * threadIdx.x + i + 1] = temp_histo[i]; 
+    } 
+  }
+
+  __syncthreads();
```

Comment:
Is the `__syncthreads()` here necessary? It looks like the `while` loop that follows doesn't have any dependencies on what happens before the sync.

Reply:
agree that this `__syncthreads()` is unnecessary. Removed in the new commit